cloud/edgecontroller/pkg/devicecontroller/controller/downstream.go

```diff
@@ -0,0 +1,742 @@
+	dc.deviceManager.Device.Store(device.Name, &CacheDevice{ObjectMeta: device.ObjectMeta, Spec: device.Spec, Status: device.Status})
+	if ok {
+		cachedDevice := value.(*CacheDevice)
+		if isDeviceUpdated(cachedDevice, device) {
+			// if node selector updated delete from old node and create in new node
+			if isNodeSelectorUpdated(cachedDevice.Spec.NodeSelector, device.Spec.NodeSelector) {
+				dc.deviceAdded(device)
+				deletedDevice := &v1alpha1.Device{ObjectMeta: cachedDevice.ObjectMeta,
+					Spec:     cachedDevice.Spec,
+					Status:   cachedDevice.Status,
+					TypeMeta: device.TypeMeta,
+				}
+				dc.deviceDeleted(deletedDevice)
+
+			} else {
+				// TODO: add an else if condition to check if ProtocolConfig/DeviceModelReference has changed, if yes whether deviceModelReference exists
+				twin := make(map[string]*types.MsgTwin)
+				addUpdatedTwins(device.Status.Twins, twin, device.ResourceVersion)
+				addDeletedTwins(cachedDevice.Status.Twins, device.Status.Twins, twin, device.ResourceVersion)
+				msg := model.NewMessage("")
+
+				resource, err := messagelayer.BuildResource(device.Spec.NodeSelector.NodeSelectorTerms[0].MatchExpressions[0].Values[0], "device/"+device.Name+"/twin/cloud_updated", "")
+				if err != nil {
+					log.LOGGER.Warnf("Built message resource failed with error: %s", err)
+					return
+				}
+				msg.BuildRouter(constants.DeviceControllerModuleName, constants.GroupTwin, resource, model.UpdateOperation)
+				content := types.DeviceTwinUpdate{Twin: twin}
+				content.EventID = uuid.NewV4().String()
+				content.Timestamp = time.Now().UnixNano() / 1e6
+				msg.Content = content
+
+				err = dc.messageLayer.Send(*msg)
+				if err != nil {
+					log.LOGGER.Errorf("Failed to send deviceTwin message %v due to error %v", msg, err)
+				}
+			}
+		}
+	} else {
+		// If device not present in device map means it is not modified and added.
+		dc.deviceAdded(device)
+	}
+}
+
+func addDeletedTwins(old []v1alpha1.Twin, new []v1alpha1.Twin, twin map[string]*types.MsgTwin, version string) {
+	opt := false
+	optional := &opt
+	for _, dtwin := range old {
+		if !ifTwinPresent(dtwin, new) {
+			expected := &types.TwinValue{}
+			expected.Value = &dtwin.Desired.Value
+			timestamp := time.Now().UnixNano() / 1e6
+
+			metadata := &types.ValueMetadata{Timestamp: timestamp}
+			expected.Metadata = metadata
+
+			// TODO: how to manage versioning ??
+			cloudVersion, err := strconv.ParseInt(version, 10, 64)
+			if err != nil {
+				log.LOGGER.Warnf("Failed to parse cloud version due to error %v", err)
+			}
+			twinVersion := &types.TwinVersion{CloudVersion: cloudVersion, EdgeVersion: 0}
+			msgTwin := &types.MsgTwin{
+				Expected:        expected,
+				Optional:        optional,
+				Metadata:        &types.TypeMetadata{Type: "deleted"},
+				ExpectedVersion: twinVersion,
+			}
+			twin[dtwin.PropertyName] = msgTwin
+		}
+	}
+}
+
+func ifTwinPresent(twin v1alpha1.Twin, new []v1alpha1.Twin) bool {
```

Comment:
new is a built-in function in go : https://golang.org/ref/spec#Allocation.
please use a different variable name.

Reply:
Done, Thanks!